File: elixir/strscan/lib/str_scanner.ex
defmodule StrScan do

  def eof?(string) do
    string == ""
  end

  def scan(string, regex) do
    results = Regex.scan(regex, string)

    result = parse_results(results)

    get_scan_result(string, result)
  end

  defp get_scan_result(string, result) do
    starts_with = String.starts_with?(string, result)

    if starts_with do
      get_new_string(string, result)
    else
      [nil, string]
    end
  end

  defp parse_results(results) do
    if results == [],
      do: nil,
      else: List.first(List.first(results))
  end

  defp get_new_string(string, match) do
    length = String.length(match)
    range = length..-1
    newString = String.slice(string, range)
    [match, newString]
  end

end

File: java/strscan/src/src/StrScanner.java
package src;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class StrScanner {
	private String string;
	
	public StrScanner(String string) {
		setString(string);
	}

	public String scan(String regex) {
		String toBeMatched = getString();
		Pattern p = Pattern.compile(regex);
		Matcher m = p.matcher(toBeMatched);
		
		boolean matchFound = !!m.find();

		if(matchFound) {
			String match = m.group(0);
			return startsWithMatch(match) ? setNewString(match) : null;
		}
		return null;
	}
	
	private boolean startsWithMatch(String match) {
		return getString().startsWith(match);
	}
	
	private String setNewString(String match) {
		int matchLength = match.length();
		String newString = getString().substring(matchLength);
		setString(newString);
		
		return match;
	}

	public boolean eof() {
		return getString().isEmpty();
	}

	public String getString() {
		return string;
	}

	private void setString(String string) {
		this.string = string;
	}

}
